src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/RedstoneIntegratorEntity.java

```diff
@@ -29,24 +40,37 @@ protected RedstoneIntegratorPeripheral createPeripheral() {
         return new RedstoneIntegratorPeripheral(this);
     }
 
-    public int getRedstoneInput(Direction direction) {
-        Objects.requireNonNull(level);
-        BlockPos neighbourPos = getBlockPos().relative(direction);
-        int power = level.getSignal(neighbourPos, direction);
-        if (power >= 15) return power;
-
-        BlockState neighbourState = level.getBlockState(neighbourPos);
-        return neighbourState.getBlock() == Blocks.REDSTONE_WIRE ? Math.max(power, neighbourState.getValue(RedStoneWireBlock.POWER)) : power;
+    /**
+     * This method is safely to be called from multiple threads at any time.
+     */
+    public int getInput(ComputerSide direction) {
+        // no need to lock because JVM promise the int will never be half-updated
+        return this.inputs.getOrDefault(direction, 0);
     }
 
-    private void setRedstoneOutput(Direction direction, int power) {
-        int old = this.power[direction.get3DDataValue()];
-        this.power[direction.get3DDataValue()] = power;
-        if (old != power) {
-            if (level != null)
-                RedstoneUtil.propagateRedstoneOutput(level, getBlockPos(), direction);
+    /**
+     * This method should only be called from main thread
+     */
+    public void setInput(ComputerSide side, int input) {
+        Integer old = this.inputs.getOrDefault(side, Integer.ZERO);
+        if (old.intValue() == input) {
+            return;
+        }
+        this.inputs.put(side, input);
+        SwarmEventDispatcher.dispatch(REDSTONE_EVENT_ID, this.getPeripheral(), side.getName());
```

Comment:
Maybe add `old` to the event? Could be a neat addition

Reply:
Can you describe more specifically what does the `old` do?

If that's the old input value, I believe the user should save them by him/herself.

Reply:
> Can you describe more specifically what does the `old` do?
> 
> If that's the old input value, I believe the user should save them by him/herself.

Yeah, the old input value

Sure, they could save them by themselves, but it would not hurt to just include it